src/framework/components/esmscript/component.js

```diff
@@ -0,0 +1,525 @@
+    }
+
+    /**
+     * A Dictionary object where each key is a string and each value is an AttributeDefinition.
+     * @typedef {Object.<string, AttributeDefinition>} AttributeDefinitionDict
+     */
+
+    /**
+     * This function recursively populates an object with attributes based on an attribute definition.
+     * Only attributes defined in the definition. Note that this does not perform any type-checking.
+     * If no attribute is specified it uses the default value from the attribute definition if available.
+     *
+     * @param {Object} object - The object to populate with attributes
+     * @param {AttributeDefinitionDict} attributeDefDict - The definition
+     * @param {Object} attributes - The attributes to apply
+     *
+     * @example
+     * const attributes = { someNum: 1, nested: { notStr: 2, ignoredValue: 20 }}
+     * const definitions = {
+     *  someNum: { type: 'number' },
+     *  nested: {
+     *      notStr: { type: 'string' },
+     *      otherValue: { type: 'number', default: 3 }
+     *  }
+     * }
+     *
+     * populateWithAttributes(object, attributeDefDict, attributes)
+     * // outputs { someNum: 1, nested: { notStr: 2, otherValue: 3 }}
+     */
+    static populateWithAttributes(object, attributeDefDict, attributes) {
+
+        for (const attributeName in attributeDefDict) {
+            const attributeDefinition = attributeDefDict[attributeName];
+
+            if (EsmScriptComponent.isValidAttributeDefinition(attributeDefinition)) {
+                if (!Object.hasOwn(object, attributeName)) {
+                    object[attributeName] = attributes?.[attributeName] || attributeDefinition.default;
+                }
+            } else if (typeof attributeDefinition === 'object') {
+
+                this.populateWithAttributes(
+                    object[attributeName] = {},
+                    attributeDefinition,
+                    attributes?.[attributeName]
+                );
+            }
+        }
+    }
+
+    /**
+     * Destroy the script instance that is attached to an entity.
+     *
+     * @param {string} moduleSpecifier - The
+     * name or type of {@link ScriptType}.
+     * @returns {boolean} If it was successfully destroyed.
+     * @example
+     * entity.script.destroy('playerController');
+     */
+    destroy(moduleSpecifier) {
+
+        const module = this.modules.get(moduleSpecifier);
+
+        if (module) {
+
+            // Remove from local data
+            this.modules.delete(moduleSpecifier);
+            this.attributeDefinitions.delete(moduleSpecifier);
+            this.modulesWithPostUpdate.delete(moduleSpecifier);
+            this.modulesWithPostUpdate.delete(moduleSpecifier);
+
+            // Fire component level events
+            this.fire('destroy', moduleSpecifier, module);
+            this.fire('destroy:' + moduleSpecifier, module);
+
+            // Call modules destroy if present
+            module.destroy?.();
+
```

Comment:
do you need to remove the listener too?
`this.system.app.assets.once(`load:url:${this.system.app.assets.prefix + moduleSpecifier}`